### PR TITLE
Better handling of URLS in Charm Editor

### PR DESCRIPTION
**Pull request by Devorein**
image
image
image

### CODE REVIEW

File: components/common/CharmEditor/CharmEditor.tsx [hunk 1]
```diff
@@ -130,6 +133,7 @@ export function charmEditorPlugins({
     // this trackPlugin should be called before the one below which calls onSelectionSet().
     // TODO: find a cleaner way to combine this logic?
     trackPlugins({ onSelectionSet, key: suggestionsPluginKey }),
+    linkPlugins({ key: linksPluginKey }),
```
Comment (mattcasey) [inline components/common/CharmEditor/CharmEditor.tsx:136]: do we also want to remove link.plugins() from the original link component?

File: components/common/CharmEditor/components/floatingMenu/LinkSubMenu.tsx [hunk 2]
```diff
@@ -1,146 +1,50 @@
...
-          }}
-          onClick={() => {
-            link.updateLink()(view.state, view.dispatch);
-          }}
-        />
-      </MenuButton>
-    </Box>
+      <FieldLabel variant='subtitle2'>Link</FieldLabel>
+      <TextField value={href} onChange={(e) => setHref(e.target.value)} autoFocus onKeyDown={handleSubmit} />
+    </Stack>
```
Comment (mattcasey) [inline components/common/CharmEditor/components/floatingMenu/LinkSubMenu.tsx:48]: Can we add input for 'title' as well? Right now I can't seem to find a way to change the text that appears for the link. Here's Notion's edit form:
image

File: components/common/CharmEditor/components/link/LinksPopup.tsx [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                  disabled={readOnly}
+                  onClick={deleteHref}
+                  sx={{
+                    py: 1
+                  }}
+                >
+                  <ListItemIcon>
+                    <DeleteOutlinedIcon fontSize='small' />
+                  </ListItemIcon>
+                  <Typography variant='subtitle1'>Delete</Typography>
```
Comment (mattcasey) [inline components/common/CharmEditor/components/link/LinksPopup.tsx:161]: Change "Delete" to "Remove link". Delete sounds like we might remove the whole text

File: components/common/CharmEditor/components/link/LinksPopup.tsx [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+      <ClickAwayListener onClickAway={hideTooltip}>
+        <Grow
+          in
+          style={{
+            transformOrigin: 'left top'
+          }}
+          easing={{
+            enter: 'ease-in-out'
+          }}
+          timeout={250}
```
Comment (mattcasey) [inline components/common/CharmEditor/components/link/LinksPopup.tsx:85]: Can we make the tooltip appearance even more like Notion:
- get rid of the animation, it's kind of distracting
- wait longer, maybe even a full second before tooltip appears
- tooltip should always appear at the same spot, below the bottom left corner of the link

File: components/common/CharmEditor/components/link/LinksPopup.tsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+              }}
+            >
+              <LanguageOutlinedIcon
+                sx={{
+                  fontSize: 14
+                }}
+                color='secondary'
+              />
+              <Typography variant='subtitle1' color='secondary'>
+                {href}
```
Comment (mattcasey) [inline components/common/CharmEditor/components/link/LinksPopup.tsx:109]: can you give this a max-width, like Notion:
image

Comment (mattcasey) [reply components/common/CharmEditor/components/link/LinksPopup.tsx:109]: We really need to make a component for this heh:
```
<Typography
        sx={{
          flexGrow: 1,
          whitespace: 'nowrap',
          overflow: 'hidden',
          textOverflow: 'ellipsis'
        }} />
        ```

File: components/common/CharmEditor/components/link/LinksPopup.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+import { useSnackbar } from 'hooks/useSnackbar';
+
+import { hideSuggestionsTooltip } from '../@bangle.dev/tooltip/suggest-tooltip';
+
+import type { LinkPluginState } from './link.plugins';
+
+export function LinksPopup({ pluginKey, readOnly }: { pluginKey: PluginKey<LinkPluginState>; readOnly: boolean }) {
+  const { showMessage } = useSnackbar();
+  const view = useEditorViewContext();
+  const [linkView, setLinkView] = useState<'link-textfield' | 'options'>('options');
```
Comment (mattcasey) [inline components/common/CharmEditor/components/link/LinksPopup.tsx:31]: The name of the views is a little confusing, 'options' sounds like an editable view.. what about "view" | "edit" ? Or even "first_view" | "edit_view"..

**Review by mattcasey — APPROVED** [review]
I submitted most of the changes I requested in a separate PR, while trying to help out and understand the hovering flow :)

File: components/common/CharmEditor/CharmEditor.tsx (hunk 1 above)
Comment (Devorein) [reply components/common/CharmEditor/CharmEditor.tsx:136]: Thanks for catching that. Yes most definitely.

File: components/common/CharmEditor/components/floatingMenu/LinkSubMenu.tsx (hunk 2 above)
Comment (Devorein) [reply components/common/CharmEditor/components/floatingMenu/LinkSubMenu.tsx:48]: Yeah lets do that. I was unsure on whether to add that one or not.

**Pull request merged**